File: src/main/kotlin/mk/ukim/finki/dnick/hosting/service/GroupService.kt
package mk.ukim.finki.dnick.hosting.service

import mk.ukim.finki.dnick.hosting.generated.model.GroupCreateRequest
import mk.ukim.finki.dnick.hosting.generated.model.GroupUpdateRequest
import mk.ukim.finki.dnick.hosting.model.entity.Application
import mk.ukim.finki.dnick.hosting.repository.ApplicationRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.server.ResponseStatusException

@Service
class GroupService(
    private val applicationRepository: ApplicationRepository,
    private val namespaceService: NamespaceService
) {

    @Transactional(readOnly = true)
    fun getGroups(pageable: Pageable): Page<Application> {
        return applicationRepository.findAll(pageable)
    }

    @Transactional
    fun createGroup(request: GroupCreateRequest): Application {
        return applicationRepository.save(
            Application(
                name = request.name,
                namespace = namespaceService.resolveUserNamespace(),
            )
        )
    }

    @Transactional
    fun updateGroup(id: Int, request: GroupUpdateRequest): Application {
        return applicationRepository.findByIdOrNull(id)?.let {
            it.name = request.name
            applicationRepository.save(it)
        } ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "Group not found")
    }

    @Transactional
    fun deleteGroup(id: Int) {
        applicationRepository.findByIdOrNull(id)?.let {
            applicationRepository.delete(it)
        } ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "Group not found")
    }

}

File: frontend/src/lib/service/groupService.ts
import { BASE_URL } from '$lib/config';
import {
  Configuration,
  GroupApi,
  type Group,
  type GroupCreateRequest,
  type GroupsResponse,
  type GroupUpdateRequest,
  type Pageable
} from '$lib/generated';

const configuration = new Configuration({
  basePath: BASE_URL
});

const groupApi = new GroupApi(configuration);

export async function getAllGroups(pageable: Pageable): Promise<GroupsResponse> {
  return await groupApi.getAllGroups({
    page: pageable.page,
    size: pageable.size,
    sort: pageable.sort
  });
}

export async function createGroup(request: GroupCreateRequest): Promise<Group> {
  return await groupApi.createGroup({
    groupCreateRequest: request
  });
}

export async function updateGroup(id: number, request: GroupUpdateRequest): Promise<Group> {
  return await groupApi.updateGroup({
    id,
    groupUpdateRequest: request
  });
}

export async function deleteGroup(id: number) {
  return await groupApi.deleteGroup({ id });
}
